network/dhtree.go

```diff
@@ -897,277 +824,159 @@ func (l *treeLabel) decode(data []byte) error {
-	if err := tmp.token.decode(data); err != nil {
-		return err
+	tmp.rootSeq = binary.BigEndian.Uint64(data[:8])
+	tmp.seq = binary.BigEndian.Uint64(data[8:16])
+	data = data[16:]
+	for len(data) > 0 {
+		var s bootstrapHopSig
+		if !wireChopSlice(s.key[:], &data) {
+			return wireDecodeError
+		} else if !wireChopSlice(s.sig[:], &data) {
+			return wireDecodeError
+		}
+		tmp.bhs = append(tmp.bhs, s)
 	}
-	*s = tmp
+	*dbs = tmp
 	return nil
 }
 
-/***************
- * dhtTeardown *
- ***************/
-
-type dhtTeardown struct {
-	seq     uint64
-	key     publicKey
-	root    publicKey
-	rootSeq uint64
-}
+/**************
+ * dhtTraffic *
+ **************/
 
-func (t *dhtTeardown) getMapKey() dhtMapKey {
-	return dhtMapKey{t.key, t.root, t.rootSeq}
+type dhtWatermark struct {
+	key publicKey
+	seq uint64
 }
 
-func (t *dhtTeardown) encode(out []byte) ([]byte, error) {
+func (m *dhtWatermark) encode(out []byte) ([]byte, error) {
+	out = append(out, m.key[:]...)
 	seq := make([]byte, 8)
-	binary.BigEndian.PutUint64(seq, t.seq)
+	binary.BigEndian.PutUint64(seq, m.seq)
 	out = append(out, seq...)
-	out = append(out, t.key[:]...)
-	out = append(out, t.root[:]...)
-	rseq := make([]byte, 8)
-	binary.BigEndian.PutUint64(rseq, t.rootSeq)
-	out = append(out, rseq...)
 	return out, nil
 }
 
-func (t *dhtTeardown) decode(data []byte) error {
-	var tmp dhtTeardown
-	if len(data) < 8 {
-		return wireDecodeError
-	}
-	tmp.seq, data = binary.BigEndian.Uint64(data[:8]), data[8:]
+func (m *dhtWatermark) decode(data []byte) error {
+	var tmp dhtWatermark
 	if !wireChopSlice(tmp.key[:], &data) {
 		return wireDecodeError
-	} else if !wireChopSlice(tmp.root[:], &data) {
-		return wireDecodeError
-	} else if len(data) != 8 {
+	}
+	if len(data) < 8 {
 		return wireDecodeError
 	}
-	tmp.rootSeq = binary.BigEndian.Uint64(data)
-	*t = tmp
+	tmp.seq = binary.BigEndian.Uint64(data[:8])
+	data = data[8:]
+	*m = tmp
 	return nil
 }
 
-/**************
- * dhtTraffic *
- **************/
+func (m *dhtWatermark) chop(ptr *[]byte) bool {
+	if ptr == nil {
+		return false
+	}
+	if err := m.decode(*ptr); err != nil {
+		return false
+	}
+	*ptr = (*ptr)[len(m.key)+8:]
+	return true
+}
 
-type dhtTraffic struct {
+type baseTraffic struct {
 	source  publicKey
 	dest    publicKey
 	kind    byte // in-band vs out-of-band, TODO? separate type?
 	payload []byte
 }
 
-func (t *dhtTraffic) encode(out []byte) ([]byte, error) {
+func (t *baseTraffic) encode(out []byte) ([]byte, error) {
 	out = append(out, t.source[:]...)
 	out = append(out, t.dest[:]...)
 	out = append(out, t.kind)
 	out = append(out, t.payload...)
 	return out, nil
 }
 
-func (t *dhtTraffic) decode(data []byte) error {
-	var tmp dhtTraffic
+func (t *baseTraffic) decode(data []byte) error {
+	var tmp baseTraffic
 	if !wireChopSlice(tmp.source[:], &data) {
 		return wireDecodeError
 	} else if !wireChopSlice(tmp.dest[:], &data) {
 		return wireDecodeError
-	}
-	if len(data) < 1 {
+	} else if len(data) < 1 {
```

Comment:
should this be `len(data) <= 1` ?

Reply:
The `len(data) < 1` is intentional here, to allow for the possibility that the payload to the packet is 0 bytes. Whether or not that's *useful* is another matter, but it would mean that a the exported `net.PacketConn.ReadFrom` returns with no error and 0 bytes read. Ygg doesn't use that for anything, but it's possible other code using ironwood could decide to use that for e.g. some kind of keepalive traffic.